File: gauss_implementation/code in c/gauss.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <windows.h>

#define MAXN 2000
int N;

float A[MAXN][MAXN], B[MAXN], X[MAXN];

void gauss();
void configurar_parametros(int argc, char **argv);
void inicializar_dados();

void configurar_parametros(int argc, char **argv) {
    if (argc >= 2) {
        N = atoi(argv[1]);
        if (N < 1 || N > MAXN) {
            exit(0);
        }
    } else {
        exit(0);
    }
}

void inicializar_dados() {
    for (int linha = 0; linha < N; linha++) {
        for (int coluna = 0; coluna < N; coluna++) {
            A[linha][coluna] = (float)rand() / 32768.0;
        }
    }
    
    for (int i = 0; i < N; i++) {
        B[i] = (float)rand() / 32768.0;
        X[i] = 0.0;
    }
}


void gauss() {
    for (int pivo = 0; pivo < N - 1; pivo++) {
        for (int linha = pivo + 1; linha < N; linha++) {
            float multiplicador = A[linha][pivo] / A[pivo][pivo];
            for (int coluna = pivo; coluna < N; coluna++) {
                A[linha][coluna] -= A[pivo][coluna] * multiplicador;
            }
            B[linha] -= B[pivo] * multiplicador;
        }
    }

    for (int linha = N - 1; linha >= 0; linha--) {
        X[linha] = B[linha];
        for (int coluna = N-1; coluna > linha; coluna--) {
            X[linha] -= A[linha][coluna] * X[coluna];
        }
        X[linha] /= A[linha][linha];
    }
}

int main(int argc, char **argv) {
    LARGE_INTEGER frequencia, inicio, fim;
    double tempo;

    QueryPerformanceFrequency(&frequencia);
    
    configurar_parametros(argc, argv);
    inicializar_dados();

    QueryPerformanceCounter(&inicio);
    gauss();
    QueryPerformanceCounter(&fim);

    tempo = (fim.QuadPart - inicio.QuadPart) * 1000.0 / frequencia.QuadPart;

    printf("%.3f", tempo);
    return 0;
}

File: gauss_implementation/code in go/gauss.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"time"
)

const MAXN = 2000

var N int
var A [][]float32
var B []float32
var X []float32

func configurarParametros() {
	if len(os.Args) >= 3 {
		n, err := strconv.Atoi(os.Args[1])
		seed, errSeed := strconv.ParseInt(os.Args[2], 10, 64)
		if err != nil || n < 1 || n > MAXN || errSeed != nil {
			os.Exit(0)
		}
		N = n
		rand.Seed(seed)
	} else {
		os.Exit(0)
	}
}

func inicializarDados() {
	A = make([][]float32, N)
	for i := 0; i < N; i++ {
		A[i] = make([]float32, N)
		for j := 0; j < N; j++ {
			A[i][j] = rand.Float32() / 32768.0
		}
	}

	B = make([]float32, N)
	X = make([]float32, N)
	for i := 0; i < N; i++ {
		B[i] = rand.Float32() / 32768.0
		X[i] = 0.0
	}
}

func gauss() {
	for pivo := 0; pivo < N-1; pivo++ {
		for linha := pivo + 1; linha < N; linha++ {
			multiplicador := A[linha][pivo] / A[pivo][pivo]
			for coluna := pivo; coluna < N; coluna++ {
				A[linha][coluna] -= A[pivo][coluna] * multiplicador
			}
			B[linha] -= B[pivo] * multiplicador
		}
	}

	for linha := N - 1; linha >= 0; linha-- {
		X[linha] = B[linha]
		for coluna := N - 1; coluna > linha; coluna-- {
			X[linha] -= A[linha][coluna] * X[coluna]
		}
		X[linha] /= A[linha][linha]
	}
}

func main() {
	configurarParametros()
	inicializarDados()

	inicio := time.Now()
	gauss()
	tempo := time.Since(inicio).Seconds() * 1000.0

	fmt.Printf("Tempo de execucao: %.3f\n", tempo)
}
